server: TestBootstrapNewStore failed

server.TestBootstrapNewStore [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2926968&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2926968&tab=artifacts#/) on master @ [588d9d7ad7618784ee23a5e5f440a9d9d36bb0d5](https://github.com/cockroachdb/cockroach/commits/588d9d7ad7618784ee23a5e5f440a9d9d36bb0d5):

Fatal error:

```
panic: unexpected Stopped processor

```
Stack: 

```
goroutine 127270 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).registerWithRangefeedRaftMuLocked(0xc00709a400, 0x52d1ae0, 0xc008940d50, 0xc0049e6cd0, 0x1, 0x8, 0xc0049e6cd8, 0x1, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:385 +0x745
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed(0xc00709a400, 0xc0070c4180, 0x532b040, 0xc00165c2a0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:232 +0x5b7
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).RangeFeed(0xc00430d000, 0xc0070c4180, 0x532b040, 0xc00165c2a0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2502 +0x11e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).RangeFeed(0xc0009930a0, 0xc0070c4180, 0x532b040, 0xc00165c2a0, 0xc00161cf68)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:216 +0xce
github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed(0xc0033f1080, 0xc0070c4180, 0x532b040, 0xc00165c2a0, 0xc00161cf68, 0x1a8973a)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:1011 +0x54
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func1(0xc006afbf30, 0x53177e0, 0xc0033f1080, 0xc0070c4180, 0x52d1a20, 0xc00311b4c0, 0xc00220ade0, 0xc00220ae40)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:573 +0xc5
created by github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:571 +0x11e
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBootstrapNewStore
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBootstrapNewStore537493561
    test_log_scope.go:74: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestBootstrapNewStore PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
/cc @spencerkimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBootstrapNewStore.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestBootstrapNewStore [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4076933&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4076933&tab=artifacts#/) on master @ [6ff6c77e7aebec2edc947c1ed99a1f7cb67e1a3c](https://github.com/cockroachdb/cockroach/commits/6ff6c77e7aebec2edc947c1ed99a1f7cb67e1a3c):

Fatal error:

```
panic: unexpected Stopped processor

```
Stack: 

```
goroutine 13226195 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).registerWithRangefeedRaftMuLocked(0xc005bf6a80, {0x4e0d198, 0xc018ca1b30}, {{0xc001a9e818, 0x1, 0x8}, {0xc001a9e820, 0x1, 0x8}}, {0x16c89d4855b9d441, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:414 +0x798
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).rangeFeedWithRangeID(0xc005bf6a80, 0x10, 0xc013914c60, {0x4e789a0, 0xc01721bae0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:237 +0x635
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:145
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).RangeFeed(0xc0139c5200, 0xc013914c60, {0x4e789a0, 0xc01721bae0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2979 +0xdc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).RangeFeed(0x0, 0xc013914c60, {0x4e789a0, 0xc01721bae0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:216 +0xe8
github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed(0x4e789a0, 0x0, {0x4e789a0, 0xc01721bae0})
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:1157 +0x38
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func1()
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:694 +0xbb
created by github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:692 +0x18f
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBootstrapNewStore
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBootstrapNewStore889239033
    test_log_scope.go:80: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBootstrapNewStore.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestBootstrapNewStore [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4252073&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4252073&tab=artifacts#/) on master @ [a5158c4c93e9dca858b38ef7e94321f0a0a2b5c5](https://github.com/cockroachdb/cockroach/commits/a5158c4c93e9dca858b38ef7e94321f0a0a2b5c5):

Fatal error:

```
panic: close of closed channel [recovered]
	panic: close of closed channel [recovered]
	panic: close of closed channel

```
Stack: 

```
goroutine 15654111 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00e301d40, {0x5032c58, 0xc017893650})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x39a9260, 0x4f62390})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00e301d40, {0x5032c58, 0xc017893980})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x39a9260, 0x4f62390})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/pebble/sstable.(*writeQueue).finish(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/sstable/write_queue.go:108
github.com/cockroachdb/pebble/sstable.(*Writer).Close(0xc00370e000)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/sstable/writer.go:1047 +0xc7
github.com/cockroachdb/cockroach/pkg/storage.(*SSTWriter).Close(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/sst_writer.go:296
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*multiSSTWriter).Close(0xc005c0b028)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_snapshot.go:221 +0x2a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*kvBatchSnapshotStrategy).Receive(_, {_, _}, {_, _}, {{0x15, 0x4, 0xc01062a540, 0xc013c58960, 0xc012002560, ...}, ...})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_snapshot.go:272 +0x6c5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).receiveSnapshot(0xc00bc60000, {0x5032c58, 0xc017893980}, 0xc00db01680, {0x7f2c83996598, 0xc00bc50090})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_snapshot.go:679 +0x5c5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleSnapshot.func1({0x5032c58, 0xc017893980})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:82 +0x145
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc00e301d40, {0x5032c58, 0xc017893980}, {0x3d59900, 0xc017893680}, 0xc003754c48)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:344 +0xdd
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleSnapshot(0xc00bc60000, {0x5032c58, 0xc017893650}, 0xc00db01680, {0x7f2c82ede968, 0xc00bc50090})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:72 +0xcc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).RaftSnapshot.func1.1({0x5095ab0, 0xc00bc50090}, 0xc0166dec80, {0x5032c58, 0xc017893650})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/raft_transport.go:434 +0x13a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).RaftSnapshot.func1({0x5032c58, 0xc017893650})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/raft_transport.go:435 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x16f
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBootstrapNewStore
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBootstrapNewStore868387766
    test_log_scope.go:80: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBootstrapNewStore.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/rangefeed/processor.go
+++ b/pkg/kv/kvserver/rangefeed/processor.go
@@ -210,15 +210,16 @@ type CatchUpIteratorConstructor func() *CatchUpIterator
 // calling its Close method when it is finished. If the iterator is nil then
 // no initialization scan will be performed and the resolved timestamp will
 // immediately be considered initialized.
-func (p *Processor) Start(stopper *stop.Stopper, rtsIterFunc IntentScannerConstructor) {
+func (p *Processor) Start(stopper *stop.Stopper, rtsIterFunc IntentScannerConstructor) error {
 	ctx := p.AnnotateCtx(context.Background())
 	if err := stopper.RunAsyncTask(ctx, "rangefeed.Processor", func(ctx context.Context) {
 		p.run(ctx, p.RangeID, rtsIterFunc, stopper)
 	}); err != nil {
-		pErr := roachpb.NewError(err)
-		p.reg.DisconnectWithErr(all, pErr)
+		p.reg.DisconnectWithErr(all, roachpb.NewError(err))
 		close(p.stoppedC)
+		return err
 	}
+	return nil
 }
 
 // run is called from Start and runs the rangefeed.

--- a/pkg/kv/kvserver/rangefeed/processor_test.go
+++ b/pkg/kv/kvserver/rangefeed/processor_test.go
@@ -133,8 +133,9 @@ func rangeFeedCheckpoint(span roachpb.Span, ts hlc.Timestamp) *roachpb.RangeFeed
 const testProcessorEventCCap = 16
 
 func newTestProcessorWithTxnPusher(
-	rtsIter storage.SimpleMVCCIterator, txnPusher TxnPusher,
+	t *testing.T, rtsIter storage.SimpleMVCCIterator, txnPusher TxnPusher,
 ) (*Processor, *stop.Stopper) {
+	t.Helper()
 	stopper := stop.NewStopper()
 
 	var pushTxnInterval, pushTxnAge time.Duration = 0, 0 // disable
@@ -153,7 +154,7 @@ func newTestProcessorWithTxnPusher(
 		EventChanCap:         testProcessorEventCCap,
 		CheckStreamsInterval: 10 * time.Millisecond,
 	})
-	p.Start(stopper, makeIntentScannerConstructor(rtsIter))
+	require.NoError(t, p.Start(stopper, makeIntentScannerConstructor(rtsIter)))
 	return p, stopper
 }
 
@@ -164,13 +165,16 @@ func makeIntentScannerConstructor(rtsIter storage.SimpleMVCCIterator) IntentScan
 	return func() IntentScanner { return NewLegacyIntentScanner(rtsIter) }
 }
 
-func newTestProcessor(rtsIter storage.SimpleMVCCIterator) (*Processor, *stop.Stopper) {
-	return newTestProcessorWithTxnPusher(rtsIter, nil /* pusher */)
+func newTestProcessor(
+	t *testing.T, rtsIter storage.SimpleMVCCIterator,
+) (*Processor, *stop.Stopper) {
+	t.Helper()
+	return newTestProcessorWithTxnPusher(t, rtsIter, nil /* pusher */)
 }
 
 func TestProcessorBasic(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	p, stopper := newTestProcessor(nil /* rtsIter */)
+	p, stopper := newTestProcessor(t, nil /* rtsIter */)
 	defer stopper.Stop(context.Background())
 
 	// Test processor without registrations.
@@ -433,13 +437,13 @@ func TestNilProcessor(t *testing.T) {
 	// to call on a nil Processor.
 	stopper := stop.NewStopper()
 	defer stopper.Stop(context.Background())
-	require.Panics(t, func() { p.Start(stopper, nil) })
+	require.Panics(t, func() { _ = p.Start(stopper, nil) })
 	require.Panics(t, func() { p.Register(roachpb.RSpan{}, hlc.Timestamp{}, nil, false, nil, nil) })
 }
 
 func TestProcessorSlowConsumer(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	p, stopper := newTestProcessor(nil /* rtsIter */)
+	p, stopper := newTestProcessor(t, nil /* rtsIter */)
 	defer stopper.Stop(context.Background())
 
 	// Add a registration.
@@ -565,7 +569,7 @@ func TestProcessorInitializeResolvedTimestamp(t *testing.T) {
 	}, nil)
 	rtsIter.block = make(chan struct{})
 
-	p, stopper := newTestProcessor(rtsIter)
+	p, stopper := newTestProcessor(t, rtsIter)
 	defer stopper.Stop(context.Background())
 
 	// The resolved timestamp should not be initialized.
@@ -727,7 +731,7 @@ func TestProcessorTxnPushAttempt(t *testing.T) {
 		return nil
 	})
 
-	p, stopper := newTestProcessorWithTxnPusher(nil /* rtsIter */, &tp)
+	p, stopper := newTestProcessorWithTxnPusher(t, nil /* rtsIter */, &tp)
 	defer stopper.Stop(context.Background())
 
 	// Add a few intents and move the closed timestamp forward.
@@ -817,7 +821,7 @@ func TestProcessorConcurrentStop(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	const trials = 10
 	for i := 0; i < trials; i++ {
-		p, stopper := newTestProcessor(nil /* rtsIter */)
+		p, stopper := newTestProcessor(t, nil /* rtsIter */)
 
 		var wg sync.WaitGroup
 		wg.Add(6)
@@ -863,7 +867,7 @@ func TestProcessorConcurrentStop(t *testing.T) {
 // observes only operations that are consumed after it has registered.
 func TestProcessorRegistrationObservesOnlyNewEvents(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	p, stopper := newTestProcessor(nil /* rtsIter */)
+	p, stopper := newTestProcessor(t, nil /* rtsIter */)
 	defer stopper.Stop(context.Background())
 
 	firstC := make(chan int64)

--- a/pkg/kv/kvserver/replica_rangefeed.go
+++ b/pkg/kv/kvserver/replica_rangefeed.go
@@ -398,7 +398,15 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 		return rangefeed.NewSeparatedIntentScanner(iter)
 	}
 
-	p.Start(r.store.Stopper(), rtsIter)
+	// NB: This only errors if the stopper is stopping, and we have to return here
+	// in that case. We do check ShouldQuiesce() below, but that's not sufficient
+	// because the stopper has two states: stopping and quiescing. If this errors
+	// due to stopping, but before it enters the quiescing state, then the select
+	// below will fall through to the panic.
+	if err := p.Start(r.store.Stopper(), rtsIter); err != nil {
+		errC <- roachpb.NewError(err)
+		return nil
+	}
 
 	// Register with the processor *before* we attach its reference to the
 	// Replica struct. This ensures that the registration is in place before

